hnn_core/viz.py

```diff
@@ -64,20 +78,47 @@ def plot_hist_input(net, ax=None, show=True):
         The matplotlib figure handle.
     """
     import matplotlib.pyplot as plt
-    spikes = np.array(sum(net.spikes.times, []))
-    gids = np.array(sum(net.spikes.gids, []))
-    valid_gids = np.r_[[v for (k, v) in net.gid_dict.items()
-                        if k.startswith('evprox')]]
-    mask_evprox = np.in1d(gids, valid_gids)
-    valid_gids = np.r_[[v for (k, v) in net.gid_dict.items()
-                        if k.startswith('evdist')]]
-    mask_evdist = np.in1d(gids, valid_gids)
-    bins = np.linspace(0, net.params['tstop'], 50)
+    spike_times = np.array(sum(spikes._times, []))
+    spike_types_data = np.array(sum(spikes._types, []))
+    default_types = ['evprox', 'evdist', 'common', 'extgauss', 'extpois']
+    cell_types = ['L5_pyramidal', 'L5_basket', 'L2_pyramidal', 'L2_basket']
+    input_types = np.setdiff1d(np.unique(spike_types_data), cell_types)
+    spike_types_mask = {s_type: np.in1d(spike_types_data, s_type)
+                        for s_type in input_types}
+
+    if isinstance(spike_types, str):
+        spike_types = {spike_types: [spike_types]}
+    elif isinstance(spike_types, list):
+        spike_types = {s_type: [s_type] for s_type in spike_types}
+    elif spike_types is None:
+        spike_types = {s_type: [s_type] for s_type in default_types}
+    elif not isinstance(spike_types, dict):
+        raise TypeError('spike_types should be str, list, dict, or None')
+
+    spike_mask = {}
+    for (s_label, s_list) in spike_types.items():
+        s_label_mask = []
+        for s_type in s_list:
+            s_mask_list = []
+            for (s_key, s_mask) in spike_types_mask.items():
+                if s_key.startswith(s_type):
+                    s_mask_list.append(s_mask)
+            s_mask_list = np.logical_or.reduce(s_mask_list)
+            s_label_mask.append(s_mask_list)
+        spike_mask[s_label] = np.logical_or.reduce(s_label_mask)
```

Comment:
let me do a pass over this. Don't touch your code locally, otherwise you'll get merge conflicts. Meanwhile you can mull over the question above about expanding the function to be more general